libcxx.mixin

```diff
@@ -0,0 +1,10 @@
+build:
+  libcxx:
+    cmake-args:
+      - "-DCMAKE_C_COMPILER=clang-6.0"
+      - "-DCMAKE_CXX_COMPILER=clang++-6.0"
+      - "-DCMAKE_CXX_FLAGS=-stdlib=libc++ -D_LIBCPP_ENABLE_THREAD_SAFETY_ANNOTATIONS"
+      - "-DFORCE_BUILD_POCO=ON"
+      - "-DFORCE_BUILD_TINYXML=ON"
```

Comment:
I don't like the concept of passing these package specific defines in the mixin. Simply because it makes the mixin less reusable / generic. I am not sure if there is any other way to enforce this or if we should reconsider how we do these vendor packages.

Reply:
Can you please elaborate on what generic applicability you want from the contributed mixins?

If your concern is that these mixins only apply to ROS 2 builds, then I think we should move them into a manifest (either in this repository or elsewhere) of ROS 2-specific colcon mixins and concentrate our efforts there, rather than trying to build a generic library of mixins which I think have the potential to become generified to the point of uselessness  where we have one mixin mapped 1:1 to a plain colcon argument (if you'll excuse the slippery slope).

I think that we need to balance their generic usefulness to any user of colcon with the fact that we have recommended using them as a utility for our own continuous integration efforts.

The package specific definitions are, per @emersonknapp, necessary for the libcxx build to succeed which I think means they should stay bundled with the other libcxx build arguments.

Reply:
> Can you please elaborate on what generic applicability you want from the contributed mixins?

That it provides a certain feature / configuration in a workspace agnostic way. You should be able to use the mixin with an workspace to be generically usable.

> have the potential to become generified to the point of uselessness where we have one mixin mapped 1:1 to a plain colcon argument (if you'll excuse the slippery slope).

As seen in this specific patch the generic part of the mixin does contain a set of setting which together form the `libcxx` mixin. So that pretty much demonstrates that it doesn't have to degenerate into a 1:1 argument mapping.

> If your concern is that these mixins only apply to ROS 2 builds, then I think we should move them into a manifest (either in this repository or elsewhere) of ROS 2-specific colcon mixins and concentrate our efforts there

Yes, I think this repo should not contain ROS 2 workspace specific information. While it would certainly be possible to just move this into a ROS 2 specific mixin repo I would still argue that that would be undesired. Simply because the mixin would still be coupled with the content of a specific mixin which from a software engineering point of view is undesired. So I think taking a step back and considering how this could be achieved in a uncoupled fashion is necessary.

Reply:
> That it provides a certain feature / configuration in a workspace agnostic way. You should be able to use the mixin with an workspace to be generically usable.

Thanks. I was not keeping this in mind when thinking about mixins since I was using them to solve a separate problem of concise duplicability between CI and local builds.

> Yes, I think this repo should not contain ROS 2 workspace specific information.

Roger.

> Simply because the mixin would still be coupled with the content of a specific mixin which from a software engineering point of view is undesired.

I'm not understanding this sentence. Could it be that you meant "Simply because the mixin would still be coupled with the content of a specific _package_ which from a software engineering point of view is undesired."?

Reply:
Yeah, it should read: the mixin would still be coupled with the content of a specific ~~mixin~~ workspace

Reply:
So, from this conversation, are you thinking we should 
* create a separate mixin repository
* Just make these settings a jenkins job configuration for the build command, and not a mixin at all
?